## Base de datos de ejemplo DVD Rental

In [ ]:
%load_ext sql

The sql extension is already loaded. To reload it, use:
  %reload_ext sql


In [ ]:
%alias dropdb docker exec -i postgres_container dropdb -U postgres 
%alias createdb docker exec -i postgres_container createdb -U postgres 
%alias psql docker exec -i postgres_container psql -U postgres 

In [ ]:
dropdb dvdrental

dropdb: error: database removal failed: ERROR:  database "dvdrental" is being accessed by other users
DETAIL:  There is 1 other session using the database.


In [ ]:
createdb dvdrental

createdb: error: database creation failed: ERROR:  database "dvdrental" already exists


In [ ]:
psql dvdrental < data/dvd_rental_dump.sql >/dev/null

ERROR:  type "mpaa_rating" already exists
ERROR:  type "year" already exists
ERROR:  function "_group_concat" already exists with same argument types
ERROR:  function "film_in_stock" already exists with same argument types
ERROR:  function "film_not_in_stock" already exists with same argument types
ERROR:  function "get_customer_balance" already exists with same argument types
ERROR:  function "inventory_held_by_customer" already exists with same argument types
ERROR:  function "inventory_in_stock" already exists with same argument types
ERROR:  function "last_day" already exists with same argument types
ERROR:  function "last_updated" already exists with same argument types
ERROR:  relation "customer_customer_id_seq" already exists
ERROR:  relation "customer" already exists
ERROR:  function "rewards_report" already exists with same argument types
ERROR:  function "group_concat" already exists with same argument types
ERROR:  relation "actor_actor_id_seq" already exists
ERROR:  relatio

In [ ]:
%sql postgresql://postgres:postgres@localhost/dvdrental

'Connected: postgres@dvdrental'

La base de datos **DVD Rental** tiene el siguiente modelo de entidad relación:

![png](images/PostgreSQL-Sample-Database.png)

La descripción del modelo se encuentra en la siguiente url:

http://www.postgresqltutorial.com/postgresql-sample-database/

In [ ]:
%%sql 

SELECT *
FROM FILM
LIMIT 10;

 * postgresql://postgres:***@localhost/dvdrental
10 rows affected.


[(133, 'Chamber Italian', 'A Fateful Reflection of a Moose And a Husband who must Overcome a Monkey in Nigeria', 2006, 1, 7, Decimal('4.99'), 117, Decimal('14.99'), 'NC-17', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers'], "'chamber':1 'fate':4 'husband':11 'italian':2 'monkey':16 'moos':8 'must':13 'nigeria':18 'overcom':14 'reflect':5"),
 (384, 'Grosse Wonderful', 'A Epic Drama of a Cat And a Explorer who must Redeem a Moose in Australia', 2006, 1, 5, Decimal('4.99'), 49, Decimal('19.99'), 'R', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Behind the Scenes'], "'australia':18 'cat':8 'drama':5 'epic':4 'explor':11 'gross':1 'moos':16 'must':13 'redeem':14 'wonder':2"),
 (8, 'Airport Pollock', 'A Epic Tale of a Moose And a Girl who must Confront a Monkey in Ancient India', 2006, 1, 6, Decimal('4.99'), 54, Decimal('15.99'), 'R', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers'], "'airport':1 'ancient':18 'confront':14 'epic':4 'girl':11 'india':19 'monkey':16 'moos':8 'must':13 'pollock':2 'tale':5"),
 (98, 'Bright Encounters', 'A Fateful Yarn of a Lumberjack And a Feminist who must Conquer a Student in A Jet Boat', 2006, 1, 4, Decimal('4.99'), 73, Decimal('12.99'), 'PG-13', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers'], "'boat':20 'bright':1 'conquer':14 'encount':2 'fate':4 'feminist':11 'jet':19 'lumberjack':8 'must':13 'student':16 'yarn':5"),
 (1, 'Academy Dinosaur', 'A Epic Drama of a Feminist And a Mad Scientist who must Battle a Teacher in The Canadian Rockies', 2006, 1, 6, Decimal('0.99'), 86, Decimal('20.99'), 'PG', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Deleted Scenes', 'Behind the Scenes'], "'academi':1 'battl':15 'canadian':20 'dinosaur':2 'drama':5 'epic':4 'feminist':8 'mad':11 'must':14 'rocki':21 'scientist':12 'teacher':17"),
 (2, 'Ace Goldfinger', 'A Astounding Epistle of a Database Administrator And a Explorer who must Find a Car in Ancient China', 2006, 1, 3, Decimal('4.99'), 48, Decimal('12.99'), 'G', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers', 'Deleted Scenes'], "'ace':1 'administr':9 'ancient':19 'astound':4 'car':17 'china':20 'databas':8 'epistl':5 'explor':12 'find':15 'goldfing':2 'must':14"),
 (3, 'Adaptation Holes', 'A Astounding Reflection of a Lumberjack And a Car who must Sink a Lumberjack in A Baloon Factory', 2006, 1, 7, Decimal('2.99'), 50, Decimal('18.99'), 'NC-17', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Trailers', 'Deleted Scenes'], "'adapt':1 'astound':4 'baloon':19 'car':11 'factori':20 'hole':2 'lumberjack':8,16 'must':13 'reflect':5 'sink':14"),
 (4, 'Affair Prejudice', 'A Fanciful Documentary of a Frisbee And a Lumberjack who must Chase a Monkey in A Shark Tank', 2006, 1, 5, Decimal('2.99'), 117, Decimal('26.99'), 'G', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Commentaries', 'Behind the Scenes'], "'affair':1 'chase':14 'documentari':5 'fanci':4 'frisbe':8 'lumberjack':11 'monkey':16 'must':13 'prejudic':2 'shark':19 'tank':20"),
 (5, 'African Egg', 'A Fast-Paced Documentary of a Pastry Chef And a Dentist who must Pursue a Forensic Psychologist in The Gulf of Mexico', 2006, 1, 6, Decimal('2.99'), 130, Decimal('22.99'), 'G', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Deleted Scenes'], "'african':1 'chef':11 'dentist':14 'documentari':7 'egg':2 'fast':5 'fast-pac':4 'forens':19 'gulf':23 'mexico':25 'must':16 'pace':6 'pastri':10 'psychologist':20 'pursu':17"),
 (6, 'Agent Truman', 'A Intrepid Panorama of a Robot And a Boy who must Escape a Sumo Wrestler in Ancient China', 2006, 1, 3, Decimal('2.99'), 169, Decimal('17.99'), 'PG', datetime.datetime(2013, 5, 26, 14, 50, 58, 951000), ['Deleted Scenes'], "'agent':1 'ancient':19 'boy':11 'china':20 'escap':14 'intrepid':4 'must':13 'panorama':5 'robot':8 'sumo':16 'truman':2 'wrestler':17")]

![image.png](attachment:03353d71-047a-436c-bba1-48f0f7bf1452.png)

### Primera Pregunta:
- Cuáles son los géneros más alquilados y lo que menos

In [ ]:
%%sql

SELECT c.name AS Genero, count(r.rental_id) AS Total_Rentas
FROM rental AS r
JOIN inventory AS inv ON r.inventory_id = inv.inventory_id
JOIN film AS fil ON fil.film_id = inv.film_id 
JOIN film_category AS filca ON filca.film_id = fil.film_id
JOIN category AS c ON c.category_id = filca.category_id
GROUP BY c.name
ORDER BY 2 desc;


 * postgresql://postgres:***@localhost/dvdrental
16 rows affected.


[('Sports', 1179),
 ('Animation', 1166),
 ('Action', 1112),
 ('Sci-Fi', 1101),
 ('Family', 1096),
 ('Drama', 1060),
 ('Documentary', 1050),
 ('Foreign', 1033),
 ('Games', 969),
 ('Children', 945),
 ('Comedy', 941),
 ('New', 940),
 ('Classics', 939),
 ('Horror', 846),
 ('Travel', 837),
 ('Music', 830)]

Uniendo las tablas de inventory, rental, film, film_category y category 
podemos conocer los generos, los agrupamos por nombre de genero
los ordenamos en orden descendente respecto a la segunda columna (la cantidad de alquileres)

Genero mas alquilado es Sport y el menos Music


### Segunda Pregunta:
- Cúales son los los géneros que más vendes y los que menos en términos de cantidad de dinero, mirad tabla de Payments

In [ ]:
%%sql

SELECT c.name AS Genero, SUM(p.amount) AS Ventas_totales
FROM payment AS p
JOIN rental AS r ON r.rental_id = p.rental_id
JOIN inventory AS inv ON inv.inventory_id = r.inventory_id
JOIN film AS fil ON fil.film_id = inv.film_id 
JOIN film_category AS filca ON filca.film_id = fil.film_id
JOIN category AS c ON c.category_id = filca.category_id
GROUP BY c.name
ORDER BY 2 desc;

 * postgresql://postgres:***@localhost/dvdrental
16 rows affected.


[('Sports', Decimal('4892.19')),
 ('Sci-Fi', Decimal('4336.01')),
 ('Animation', Decimal('4245.31')),
 ('Drama', Decimal('4118.46')),
 ('Comedy', Decimal('4002.48')),
 ('New', Decimal('3966.38')),
 ('Action', Decimal('3951.84')),
 ('Foreign', Decimal('3934.47')),
 ('Games', Decimal('3922.18')),
 ('Family', Decimal('3830.15')),
 ('Documentary', Decimal('3749.65')),
 ('Horror', Decimal('3401.27')),
 ('Classics', Decimal('3353.38')),
 ('Children', Decimal('3309.39')),
 ('Travel', Decimal('3227.36')),
 ('Music', Decimal('3071.52'))]

Uniendo las tablas de payments, inventory, rental, film, film_category y category 
podemos conocer los generos, los agrupamos por nombre de genero
y sumamos los pagos, los ordenamos en orden descendente respecto a la segunda columna (la cantidad de dinero)

Generos que mas vende Sport y Sci-Fi  y los que menos menos Music y Travel


### Tercera Pregunta:
- Cuantos usuarios únicos han alquilado cada género?

In [ ]:
%%sql

SELECT c.name AS Genero, COUNT(DISTINCT cu.customer_id) AS Usuarios_pG
FROM customer AS cu
JOIN rental AS r ON r.customer_id = cu.customer_id
JOIN inventory AS inv ON inv.inventory_id = r.inventory_id
JOIN film AS fil ON fil.film_id = inv.film_id 
JOIN film_category AS filca ON filca.film_id = fil.film_id
JOIN category AS c ON c.category_id = filca.category_id
GROUP BY c.name
ORDER BY 2 desc;

 * postgresql://postgres:***@localhost/dvdrental
16 rows affected.


[('Sports', 519),
 ('Action', 510),
 ('Sci-Fi', 507),
 ('Family', 501),
 ('Drama', 501),
 ('Animation', 500),
 ('Comedy', 495),
 ('Foreign', 493),
 ('Documentary', 483),
 ('Children', 482),
 ('Games', 474),
 ('New', 468),
 ('Classics', 468),
 ('Horror', 451),
 ('Music', 447),
 ('Travel', 442)]

Uniendo las tablas de customer, inventory, rental, film, film_category y category 
podemos conocer los generos, los agrupamos por nombre de genero. Para identificar usuarios unicos usamos DISTINCT, los ordenamos en orden descendente respecto a la segunda columna (la suma de usuarios unicos que han rentado ese genero)

Generos que mas alquilados por usuarios unicos Sports - Action y los menoos travel y music

### Cuarta Pregunta:
- Cuanto es el gasto medio en alquiler por cada cliente?

In [ ]:
%%sql

SELECT cu.customer_id, cu.first_name AS Nombre, cu.last_name AS Apellido,
ROUND(AVG(fil.rental_rate),2) AS Gasto_Medio
FROM film AS fil 
JOIN inventory AS inv ON inv.film_id = fil.film_id 
JOIN rental AS r ON inv.inventory_id = r.inventory_id
JOIN customer AS cu ON r.customer_id = cu.customer_id
GROUP BY cu.customer_id
ORDER BY 4 desc
LIMIT 10;

 * postgresql://postgres:***@localhost/dvdrental
10 rows affected.


[(490, 'Sam', 'Mcduffie', Decimal('3.82')),
 (55, 'Doris', 'Reed', Decimal('3.81')),
 (493, 'Brent', 'Harkins', Decimal('3.77')),
 (477, 'Dan', 'Paine', Decimal('3.72')),
 (365, 'Bruce', 'Schwarz', Decimal('3.69')),
 (310, 'Daniel', 'Cabral', Decimal('3.69')),
 (43, 'Christine', 'Roberts', Decimal('3.66')),
 (347, 'Ryan', 'Salisbury', Decimal('3.66')),
 (22, 'Laura', 'Rodriguez', Decimal('3.63')),
 (433, 'Don', 'Bone', Decimal('3.63'))]

Uniendo las tablas de customer, inventory, rental, film podemos conocer los pagos por film, los agrupamos por usuarios. Para identificar usuarios unicos usamos DISTINCT, los ordenamos en orden descendente respecto a la segunda columna (Quien gasta mas o menos en promedio).Para simplificar la tabla podemmos usar ROUND, de esta forma  

Porque son demasiados reusltados he limitado el query a solo a 10 resultado, ordenados por gasto promedio Sam, Mcduffie mayor gasto -> 3.82 y menor Leona, Obrien -> 1.85 

### Quinta Pregunta
- Ordenar los países por los que más alquilan y más clientes tienen

In [ ]:
%%sql

SELECT cou.country AS pais, COUNT(DISTINCT cu.customer_id) AS Clientes,
COUNT(rental_id) AS total_alquiler
FROM customer AS cu
JOIN rental AS r ON cu.customer_id = r.customer_id
JOIN address AS add ON cu.address_id = add.address_id
JOIN city AS ci ON add.city_id = ci.city_id
JOIN country AS cou ON ci.country_id = cou.country_id
GROUP BY cou.country
ORDER BY 3 desc, 2 desc;

 * postgresql://postgres:***@localhost/dvdrental
108 rows affected.


[('India', 60, 1572),
 ('China', 53, 1426),
 ('United States', 36, 968),
 ('Japan', 31, 825),
 ('Mexico', 30, 796),
 ('Brazil', 28, 748),
 ('Russian Federation', 28, 713),
 ('Philippines', 20, 568),
 ('Turkey', 15, 388),
 ('Indonesia', 14, 367),
 ('Nigeria', 13, 352),
 ('Argentina', 13, 352),
 ('Taiwan', 10, 305),
 ('South Africa', 11, 285),
 ('Iran', 8, 225),
 ('United Kingdom', 9, 219),
 ('Poland', 8, 203),
 ('Germany', 7, 196),
 ('Italy', 7, 189),
 ('Vietnam', 6, 172),
 ('Venezuela', 7, 170),
 ('Egypt', 6, 161),
 ('Colombia', 6, 159),
 ('Ukraine', 6, 158),
 ('Spain', 5, 142),
 ('Canada', 5, 137),
 ('South Korea', 5, 135),
 ('Netherlands', 5, 134),
 ('Pakistan', 5, 128),
 ('Saudi Arabia', 5, 121),
 ('Yemen', 4, 117),
 ('Peru', 4, 112),
 ('Israel', 4, 99),
 ('France', 4, 96),
 ('Thailand', 3, 96),
 ('Bangladesh', 3, 95),
 ('Algeria', 3, 90),
 ('Ecuador', 3, 87),
 ('Malaysia', 3, 85),
 ('United Arab Emirates', 3, 84),
 ('Tanzania', 3, 83),
 ('Mozambique', 3, 80),
 ('Austria', 3, 78),
 ('Dominican Republic', 3, 77),
 ('Morocco', 3, 71),
 ('Chile', 3, 71),
 ('Belarus', 2, 66),
 ('Paraguay', 3, 62),
 ('Romania', 2, 62),
 ('Switzerland', 3, 61),
 ('Puerto Rico', 2, 61),
 ('Latvia', 2, 60),
 ('Yugoslavia', 2, 57),
 ('Azerbaijan', 2, 57),
 ('Cameroon', 2, 54),
 ('French Polynesia', 2, 54),
 ('Greece', 2, 54),
 ('Kenya', 2, 54),
 ('Sudan', 2, 54),
 ('Cambodia', 2, 53),
 ('Kazakstan', 2, 52),
 ('Angola', 2, 52),
 ('Myanmar', 2, 52),
 ('Oman', 2, 50),
 ('Bulgaria', 2, 50),
 ('Congo, The Democratic Republic of the', 2, 50),
 ('Bolivia', 2, 47),
 ('Runion', 1, 46),
 ('Anguilla', 1, 35),
 ('Brunei', 1, 35),
 ('Moldova', 1, 34),
 ('Sweden', 1, 34),
 ('Holy See (Vatican City State)', 1, 34),
 ('Greenland', 1, 34),
 ('Zambia', 1, 33),
 ('Chad', 1, 32),
 ('Virgin Islands, U.S.', 1, 32),
 ('Nauru', 1, 31),
 ('North Korea', 1, 31),
 ('Hong Kong', 1, 30),
 ('Gambia', 1, 30),
 ('Sri Lanka', 1, 30),
 ('Estonia', 1, 30),
 ('Hungary', 1, 29),
 ('Czech Republic', 1, 29),
 ('Malawi', 1, 28),
 ('Faroe Islands', 1, 28),
 ('Liechtenstein', 1, 28),
 ('Iraq', 1, 27),
 ('Turkmenistan', 1, 27),
 ('Slovakia', 1, 26),
 ('Tuvalu', 1, 26),
 ('Kuwait', 1, 26),
 ('Finland', 1, 26),
 ('Senegal', 1, 25),
 ('Bahrain', 1, 25),
 ('Armenia', 1, 25),
 ('Saint Vincent and the Grenadines', 1, 25),
 ('Lithuania', 1, 24),
 ('New Zealand', 1, 24),
 ('Tunisia', 1, 23),
 ('Ethiopia', 1, 23),
 ('French Guiana', 1, 22),
 ('Nepal', 1, 22),
 ('Madagascar', 1, 22),
 ('American Samoa', 1, 20),
 ('Tonga', 1, 18),
 ('Afghanistan', 1, 18)]

Uniendo las tablas de customer, rental, address, city, country  podemos conocer los paises y cantidad de rentas, los agrupamos por pais.De los paises el que mas cliente y rentas tiene es India
Los ordenamos de mayor a menor en numero de rentas y luego en numero de clientes usando Order by desc

### Sexta Pregunta:
- El top 10 de clientes que más alquilan 

In [ ]:
%%sql

SELECT cu.customer_id, cu.first_name AS Nombre, cu.last_name AS Apellido,
COUNT(r.rental_id) AS total_alquiler
FROM film AS fil 
JOIN inventory AS inv ON inv.film_id = fil.film_id 
JOIN rental AS r ON inv.inventory_id = r.inventory_id
JOIN customer AS cu ON r.customer_id = cu.customer_id
GROUP BY cu.customer_id
ORDER BY 4 desc
LIMIT 10;

 * postgresql://postgres:***@localhost/dvdrental
10 rows affected.


[(148, 'Eleanor', 'Hunt', 46),
 (526, 'Karl', 'Seal', 45),
 (144, 'Clara', 'Shaw', 42),
 (236, 'Marcia', 'Dean', 42),
 (75, 'Tammy', 'Sanders', 41),
 (469, 'Wesley', 'Bull', 40),
 (197, 'Sue', 'Peters', 40),
 (468, 'Tim', 'Cary', 39),
 (137, 'Rhonda', 'Kennedy', 39),
 (178, 'Marion', 'Snyder', 39)]

Uniendo las tablas de customer, inventory, rental, film podemos conocer las rentas, los agrupamos por usuarios. Para identificar usuarios que alquilaron mas, los ordenamos usando order by

Como necesitamos ver los primeros 10, limitaremos la tabla a 10 con limit

### Séptima Pregunta:
- Conteo de películas que han sido devueltas más tarde.(Pista: 
  tabla Rental, mirar la diferencia entre return_date y rental_date, y comparar con rental_duration de la tabla Film)

In [ ]:
%%sql

SELECT CASE 
WHEN rental_duration < DATE_PART('day', return_date - rental_date)THEN 'Devolucion tarde'
ELSE 'Devolucion en tiempo o processo'
END AS devolucion,
COUNT(*) AS numero_total
FROM film AS f
JOIN inventory AS inv ON f.film_id = inv.film_id
JOIN rental AS r ON inv.inventory_id = r.inventory_id
GROUP BY devolucion
ORDER BY 1 desc 

 * postgresql://postgres:***@localhost/dvdrental
2 rows affected.


[('Devolucion tarde', 6403), ('Devolucion en tiempo o processo', 9641)]

Con las tablas film, rental e inventry sacaremos la informacion
de el dia de alquiler, dia de retorno y el tiempo de alquiler establecido.
Usando la funcion DATE_PART() podemos calcular el tiempo en que el usuario tuvo el film.
Selecionamos la unidad 'day', ya que el tiempo de retorno esta estipulado en esa unidad.
Aplicaremos un CASE para distinguir de films que han sido devueltas mas tarde de lo esperado.
Podemos ver un total de 6403 peliculas devueltas tarde